fix pull image error from multiple ACRs using azure managed identity

andyzhangx opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
fix pull image error from multiple ACRs using azure managed identity

This PR continues with PR(https://github.com/kubernetes/kubernetes/pull/92330), fixed the pull image error with azure managed identity.

This PR is using per-registry cache (also with cache expiration), pls note that per-registry cache is only for managed identity, and for service principal(account name & password), it's not necessary since password is supposed to be the same in whole kubelet life time.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #92326

**Special notes for your reviewer**:
 - impact of this issue, you may hit this issue when pulling images from multiple different ACRs, and after a few more retries, it should work:
```
Events:
  Type     Reason     Age                 From                                        Message
  ----     ------     ----                ----                                        -------
  Normal   Scheduled  104s                default-scheduler                           Successfully assigned default/nginx2 to aks-nodepool1-43715599-vmss000001
  Warning  Failed     55s (x3 over 97s)   kubelet, aks-nodepool1-43715599-vmss000001  Failed to pull image "andyacr3.azurecr.io/nginx:v1": rpc error: code = Unknown desc = Error response from daemon: Get https://andyacr3.azurecr.io/v2/nginx/manifests/v1: unauthorized: authentication required, visit https://aka.ms/acr/authorization for more information.
  Warning  Failed     55s (x3 over 97s)   kubelet, aks-nodepool1-43715599-vmss000001  Error: ErrImagePull
  Normal   BackOff    27s (x4 over 97s)   kubelet, aks-nodepool1-43715599-vmss000001  Back-off pulling image "andyacr3.azurecr.io/nginx:v1"
  Warning  Failed     27s (x4 over 97s)   kubelet, aks-nodepool1-43715599-vmss000001  Error: ImagePullBackOff
  Normal   Pulling    12s (x4 over 103s)  kubelet, aks-nodepool1-43715599-vmss000001  Pulling image "andyacr3.azurecr.io/nginx:v1"
  Normal   Pulled     12s                 kubelet, aks-nodepool1-43715599-vmss000001  Successfully pulled image "andyacr3.azurecr.io/nginx:v1" in 383.922005ms
  Normal   Created    12s                 kubelet, aks-nodepool1-43715599-vmss000001  Created container nginx
  Normal   Started    12s                 kubelet, aks-nodepool1-43715599-vmss000001  Started container nginx
```

With this fix, there is no authentication error when pulling images from two ACRs using managed identity:
```
Events:
  Type    Reason     Age   From                                        Message
  ----    ------     ----  ----                                        -------
  Normal  Scheduled  14s   default-scheduler                           Successfully assigned default/nginx to aks-nodepool1-43715599-vmss000000
  Normal  Pulling    13s   kubelet, aks-nodepool1-43715599-vmss000000  Pulling image "andyacr2.azurecr.io/nginx:v1"
  Normal  Pulled     6s    kubelet, aks-nodepool1-43715599-vmss000000  Successfully pulled image "andyacr2.azurecr.io/nginx:v1" in 7.513394748s
  Normal  Created    2s    kubelet, aks-nodepool1-43715599-vmss000000  Created container nginx
  Normal  Started    2s    kubelet, aks-nodepool1-43715599-vmss000000  Started container nginx

Events:
  Type    Reason     Age   From                                        Message
  ----    ------     ----  ----                                        -------
  Normal  Scheduled  13s   default-scheduler                           Successfully assigned default/nginx2 to aks-nodepool1-43715599-vmss000000
  Normal  Pulling    13s   kubelet, aks-nodepool1-43715599-vmss000000  Pulling image "andyacr3.azurecr.io/nginx:v1"
  Normal  Pulled     2s    kubelet, aks-nodepool1-43715599-vmss000000  Successfully pulled image "andyacr3.azurecr.io/nginx:v1" in 10.224016565s
  Normal  Created    2s    kubelet, aks-nodepool1-43715599-vmss000000  Created container nginx
  Normal  Started    2s    kubelet, aks-nodepool1-43715599-vmss000000  Started container nginx
```

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
fix pull image error from multiple ACRs using azure managed identity
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
fix pull image error from multiple ACRs using azure managed identity
```

/priority important-soon
/sig cloud-provider
/area provider/azure
/triage accepted
/assign @feiskyer

andyzhangx committed:
fix pull image error from multiple ACRs using azure managed identity

fix comments

fix comment

fix comments

fix comments

fix comments

fix comments

fix bazel

andyzhangx commented:
/kind bug

andyzhangx commented:
/test pull-kubernetes-bazel-test

feiskyer commented:
/retest

andyzhangx commented:
/hold cancel
test pass: With this fix, there is no authentication error when pulling images from two ACRs using managed identity

andyzhangx commented:
/retest

liggitt commented on pkg/credentialprovider/azure/azure_credentials.go:
```diff
@@ -186,6 +191,30 @@ func (a *acrProvider) Enabled() bool {
+
+			// check whether loginServer is in cache
+			for k := range a.cacheDockerConfig {
+				if strings.EqualFold(loginServer, k) {
+					return a.cacheDockerConfig
+				}
```
We shouldn't add logic to this plugin that could not be expressed within the API of the externalized credential provider plugins, as that will prevent moving this credential provider out of tree.

Can you give a concrete example of what the `image` string will be, and what the `loginServer` string will be?

It seems like this plugin is fetching credentials based on the image and wants to cache them for any image using the same registry. Is that the case?

andyzhangx replied:
@liggitt below is an example, with Azure Managed Identity (no password), the credential is based on `loginServer`, and different `loginServer` has different credentials.
 - `image`: `andyacr3.azurecr.io/nginx:v1`
 - `loginServer`: `andyacr3.azurecr.io`

AWS also uses same implementation today:
https://github.com/kubernetes/kubernetes/blob/4261200724cd0541b4ff02cc56be866b98142823/pkg/credentialprovider/aws/aws_credentials.go#L85-L98
I think externalized credential provider plugin should also supports such logic.

Anyway, this is now a critical bug in Azure side, we need to fix it ASAP, thanks for understanding, and will check how to bring this logic to externalized credential provider plugin later.

liggitt replied:
Does the login server always match the registry portion of the image?

andyzhangx replied:
depends on image itself, if it's Azure Container Registry image, e.g. `andyacr3.azurecr.io/nginx:v1`, then it would always match since we only cache Azure Container Registry login server.

andyzhangx commented:
/retest

andyzhangx commented:
/retest

andyzhangx commented:
/retest

andyzhangx commented:
/test pull-kubernetes-e2e-aks-engine-azure

andrewsykim commented:
> We shouldn't add logic to this plugin that could not be expressed within the API of the externalized credential provider plugins, as that will prevent moving this credential provider out of tree.

I agree with @liggitt that we shouldn't add to the existing credentials providers, especially if the externalized API will handle this exact use-case (cachinig credentials on a per-registry basis).

andrewsykim commented on pkg/credentialprovider/azure/azure_credentials.go:
```diff
@@ -186,6 +191,30 @@ func (a *acrProvider) Enabled() bool {
+		if a.config != nil && a.config.UseManagedIdentityExtension {
+			loginServer := a.parseACRLoginServerFromImage(image)
+			if loginServer == "" {
+				return a.cacheDockerConfig
```
This sort of behavior is not something the externalized API will cover, if we can't determine the registry (login server), we should log an error and return an empty `credntialprovider.DockerConfig`

andrewsykim replied:
This would also result in the ACR provider returning cached docker config for non-ACR repos?

andyzhangx replied:
fixed in line 203, PTAL.

andrewsykim commented on pkg/credentialprovider/azure/azure_credentials.go:
```diff
@@ -186,6 +191,30 @@ func (a *acrProvider) Enabled() bool {
+					return a.cacheDockerConfig
+				}
+			}
+			// continue with get ARM token next since loginServer not found in cache
```
Since we removed CachingDockerConfig, the code path to fetch ARM token would run for any non-ACR repo where the loginServer is never a match. Is that the case here?

andyzhangx replied:
that's not correct, see Line 229: `klog.V(4).Infof("image(%s) is not from ACR, skip MSI authentication", image)`, if it's non-ACR repo, it won't fetch ARM token

andyzhangx replied:
anyway, return empty DockerConfig for non-ACR repo looks reasonable, will fix it soon, thanks.

andyzhangx replied:
fixed in line 202, PTAL.

andyzhangx commented:
I refactored the logic in `Provide` func in Azure provider a little, pulling 3 images from two ACRs and dockerhub successfully, PTAL. Thanks.
```
Events:
  Type    Reason     Age   From                                        Message
  ----    ------     ----  ----                                        -------
  Normal  Scheduled  25s   default-scheduler                           Successfully assigned default/nginx to aks-nodepool1-43715599-vmss000002
  Normal  Pulling    24s   kubelet, aks-nodepool1-43715599-vmss000002  Pulling image "andyacr2.azurecr.io/nginx:v1"
  Normal  Pulled     16s   kubelet, aks-nodepool1-43715599-vmss000002  Successfully pulled image "andyacr2.azurecr.io/nginx:v1" in 8.3201804s
  Normal  Created    12s   kubelet, aks-nodepool1-43715599-vmss000002  Created container nginx
  Normal  Started    11s   kubelet, aks-nodepool1-43715599-vmss000002  Started container nginx
Events:
  Type    Reason     Age   From                                        Message
  ----    ------     ----  ----                                        -------
  Normal  Scheduled  25s   default-scheduler                           Successfully assigned default/nginx to aks-nodepool1-43715599-vmss000002
  Normal  Pulling    24s   kubelet, aks-nodepool1-43715599-vmss000002  Pulling image "andyacr2.azurecr.io/nginx:v1"
  Normal  Pulled     16s   kubelet, aks-nodepool1-43715599-vmss000002  Successfully pulled image "andyacr2.azurecr.io/nginx:v1" in 8.3201804s
  Normal  Created    12s   kubelet, aks-nodepool1-43715599-vmss000002  Created container nginx
  Normal  Started    11s   kubelet, aks-nodepool1-43715599-vmss000002  Started container nginx
Events:
  Type    Reason     Age   From                                        Message
  ----    ------     ----  ----                                        -------
  Normal  Scheduled  24s   default-scheduler                           Successfully assigned default/nginx3 to aks-nodepool1-43715599-vmss000002
  Normal  Pulling    23s   kubelet, aks-nodepool1-43715599-vmss000002  Pulling image "nginx"
  Normal  Pulled     12s   kubelet, aks-nodepool1-43715599-vmss000002  Successfully pulled image "nginx" in 11.0415041s
  Normal  Created    12s   kubelet, aks-nodepool1-43715599-vmss000002  Created container nginx
  Normal  Started    12s   kubelet, aks-nodepool1-43715599-vmss000002  Started container nginx
```

feiskyer commented:
/milestone v1.20

andrewsykim commented on pkg/credentialprovider/azure/azure_credentials.go:
```diff
@@ -186,24 +191,45 @@ func (a *acrProvider) Enabled() bool {
+			// check whether loginServer is in cache
+			for k := range a.cacheDockerConfig {
+				if strings.EqualFold(loginServer, k) {
+					return a.cacheDockerConfig
```
I still don't think this does what we want. If there are two subsequent calls to `Provide` with different login servers `foo1.azurecr.io` and `foo2.azurecr.io`, the second call will update `cacheDockerConfig` and invalidate caching for foo1.

liggitt replied:
Agreed. Since this provider is caching on the registry level, its implementation should reflect that. This is essentially the request made in https://github.com/kubernetes/kubernetes/pull/92818#issuecomment-656261189 (to make the cache coherent with the significant portion of the image input).

It looks like the registry is the significant portion of the image input that affects the computed results, so the cache should be keyed by the registry when UseManagedIdentityExtension is enabled:

```
// cache credentials globally
cacheKey := ""
if a.config != nil && a.config.UseManagedIdentityExtension {
  // cache credentials at the registry level
  cacheKey = strings.Split(image, "/")[0]
}

cachedCredentials, found := cache.getUnexpiredCredentials(cacheKey)
if found {
  return cachedCredentials
}

... get credentials and store in cache under cacheKey ...
```

andyzhangx replied:
@andrewsykim thanks, I have fixed it, see L213-L215

andyzhangx replied:
@andrewsykim @liggitt could you take a look again? thanks
here `loginServer` is actually `cacheKey`, it's actually identical though with a little different implementation.

liggitt replied:
> here `loginServer` is actually `cacheKey`, it's actually identical though with a little different implementation.
> `cacheDockerConfig[loginServer] = credential`

It would be helpful to make it clear that this logic can fit within the externalized credential provider API to compute the cache key the same way the provider will (taking the registry portion of the image, and not adding in case-insensitive comparisons)

andyzhangx replied:
I have removed the case-insensitive comparisons, should fit within externalized credential provider API now. Let's merge this PR to get in-tree code work first, thanks.  @liggitt

andyzhangx commented:
/retest

andyzhangx commented:
/test pull-kubernetes-e2e-aks-engine-azure

andyzhangx commented on pkg/credentialprovider/azure/azure_credentials.go:
```diff
@@ -186,24 +191,46 @@ func (a *acrProvider) Enabled() bool {
+			}
+			// continue with get ARM token next since loginServer not found in cache
+			if cred, err := getACRDockerEntryFromARMToken(a, loginServer); err == nil {
+				a.cacheDockerConfig[loginServer] = *cred
```
fixed.

liggitt replied:
this is still modifying a value that was previously returned to callers (line 273-274 assign the returned value to cacheDockerConfig), so this is still a data race

andyzhangx commented:
@liggitt I changed the logic a little per comments, removed the `case-insensitive comparisons`, it does not matter if it's case-sensitive, try get one more token is ok.

andyzhangx commented:
/retest

andyzhangx commented:
/retest

liggitt commented on pkg/credentialprovider/azure/azure_credentials.go:
```diff
@@ -237,10 +267,10 @@ func (a *acrProvider) Provide(image string) credentialprovider.DockerConfig {
-		cfg["*.azurecr.*"] = defaultConfigEntry
 	}
+
+	a.expiration = time.Now().Add(a.lifetime)
```
This is lengthening the expiration time on existing credentials copied from any previous cacheDockerConfig (line 203). Do the acquired tokens expire? if so, that means requests for images in new registries can extend the cache expiration even as individual items in cacheDockerConfig are no longer valid.

liggitt replied:
Having a single cache expiration and cached credential represent data for distinct registries and with distinct expirations is very hard to follow and error prone. Keying the cache by the registry of the incoming image and managing cached credentials+expirations per-registry would make more sense.

andyzhangx replied:
thanks @liggitt I have updated this PR using per-registry cache (also with cache expiration), pls note that per-registry cache is only for managed identity, and for service principal(account name & password), it's not necessary since password is supposed to be the same in whole kubelet life time.

liggitt replied:
that looks much better, thanks. one last comment on how we're adding in anonymous credentials to avoid a data race

andyzhangx commented:
/retest

andyzhangx commented:
/retest

andyzhangx commented:
/retest

andyzhangx commented:
/test pull-kubernetes-e2e-azure-disk-windows

feiskyer reviewed (commented):
/lgtm
/hold (in case @liggitt may want to review again)

feiskyer commented:
/retest
(though azuredisk CSI test failures are not related to this PR's changes)

andyzhangx commented:
/retest

liggitt commented on pkg/credentialprovider/azure/azure_credentials.go:
```diff
@@ -237,10 +296,15 @@ func (a *acrProvider) Provide(image string) credentialprovider.DockerConfig {
 
-		// add ACR anonymous repo support: use empty username and password for anonymous access
-		cfg["*.azurecr.*"] = defaultConfigEntry
+	// add ACR anonymous repo support: use empty username and password for anonymous access
+	defaultConfigEntry := credentialprovider.DockerConfigEntry{
+		Username: "",
+		Password: "",
+		Email:    dummyRegistryEmail,
 	}
+	cfg["*.azurecr.*"] = defaultConfigEntry
```
in the case where we're returning a cache entry, this is modifying the shared cache object that was previously returned and can cause a data race.

Suggest making an `addAnonymousCredentials(cfg)` method and calling it from `getFromACR` when we're constructing a new cached credential set, and from the else block in Provide() when we're not using UseManagedIdentityExtension. That ensures we never modify a config after we've returned it from Provide()

andyzhangx replied:
hi @liggitt there is no shared cache object anymore, `cfg` is a newly constructed object every time in `Provide()` func, see line 253, I have removed the `cacheDockerConfig` field in this PR.

liggitt replied:
ah, I see, thanks.

liggitt reviewed (commented):
/lgtm
/approve

andyzhangx commented:
/hold cancel
@liggitt  thanks for the review. Really appreciate that!

andyzhangx commented:
/hold cancel

porrascarlos802018 commented:
Please update us on this
https://github.com/kubernetes/kubernetes/pull/92818
says will be addressed on   #96355  however , this one at the end says on hold. 

Please advise on ETA and actions required to be executed on existing clusters with this MSI problem.

andyzhangx commented:
> Please update us on this
> #92818
> says will be addressed on #96355 however , this one at the end says on hold.
> 
> Please advise on ETA and actions required to be executed on existing clusters with this MSI problem.

@porrascarlos802018 this "pull image auth error issue from **multiple**(>1) ACRs using azure managed identity" has been merged in 1.17.14, 1.18.12, 1.19.4, for existing clusters with MSI, wait for a few more tries or using single ACR could workaround
